### PR TITLE
Do not ignore the return value of `realpath`

### PR DESCRIPTION
**Pull request by siyuan0322**

### CODE REVIEW

File: modules/io/io/io_factory.cc [hunk 1]
```diff
@@ -61,7 +61,11 @@ std::unique_ptr<IIOAdaptor> IOFactory::CreateIOAdaptor(
     if (!s.ok()) {  // Assume it's a local file
       // defaulting to local file system: resolve to abs path first
       char resolved_path[PATH_MAX];
-      realpath(location_to_parse.c_str(), resolved_path);
+      char* res = realpath(location_to_parse.c_str(), resolved_path);
+      if (!res) {
+        LOG(ERROR) << "Failed to resolve realpath of " << location_to_parse;
+        return nullptr;
+      }
```
Comment (sighingnow) [inline modules/io/io/io_factory.cc:68]: It is expected. When `location_to_parse` doesn't exist, it returns failure but `resolved_path` will be still filled as the expanded path.

The open mode might be `w` and the file doesn't exist, it still returns fail, but it is valid.

Ignore the return value of `realpath` is OK here and we just delegates the failure reporting the the IO adaptor itself.

**Review by sighingnow — CHANGES_REQUESTED** [review]

File: modules/io/io/io_factory.cc (hunk 1 above)
Comment (siyuan0322) [reply modules/io/io/io_factory.cc:68]: Then I changed it to log an warning. The ignoring behavior is treated as an warning in gcc-9, and with -Werror enabled it's an error.

Comment (sighingnow) [reply modules/io/io/io_factory.cc:68]: If you want to check if it resolves successfully, you need to check if `resolved_path` is empty after resolve. It it is empty, you could raise a the warning.

**Review by sighingnow — CHANGES_REQUESTED** [review]

File: modules/io/io/io_factory.cc (hunk 1 above)
Comment (siyuan0322) [reply modules/io/io/io_factory.cc:68]: If it failed the return value will be NULL.

Comment (sighingnow) [reply modules/io/io/io_factory.cc:68]: For `w`, it fails but `resolved_path` will be filled successfully. The we SHOULDN'T raise this warning.

Instead, you need to check if `resolved_path` is null (is empty string).

Comment (siyuan0322) [reply modules/io/io/io_factory.cc:68]: I don't want to check resolved_path... The point is that I want to make use of the return value, you can treat the VLOG(2) as a debug message anyway :)

**Review by sighingnow — APPROVED** [review]

**Pull request merged**